src/test_rand_types.rs

```diff
@@ -0,0 +1,324 @@
+    let outer_size = rng.gen_range(0, ALLOC_LIMIT);
+    let mut outer_vec = Vec::with_capacity(outer_size);
+    for _ in 0..outer_size {
+        let inner_size = rng.gen_range(0, ALLOC_LIMIT);
+        let mut inner_vec = Vec::with_capacity(inner_size);
+        for _ in 0..inner_size {
+            inner_vec.push(rng.gen());
+        }
+        outer_vec.push(inner_vec);
+    }
+    outer_vec
+}
+
+fn gen_vec_hash<R: Rng + ?Sized, H: Hash>(rng: &mut R) -> Vec<H> {
+    let size = rng.gen_range(0, ALLOC_LIMIT);
+    let mut vec = Vec::with_capacity(size);
+    for _ in 0..size {
+        vec.push(gen_hash(rng))
+    }
+    vec
+}
+
+fn gen_hash<R: Rng + ?Sized, H: Hash>(rng: &mut R) -> H {
+    let mut hash_vec = vec![0u8; H::LEN];
+    rng.fill_bytes(&mut hash_vec[..]);
+    H::from_slice(&hash_vec).unwrap() // It's the right length. should never fail.
+}
+
+impl Distribution<ServiceFlags> for Standard {
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> ServiceFlags {
+        ServiceFlags::from(rng.next_u64())
+    }
+}
+
+impl Distribution<InvType> for Standard {
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> InvType {
+        // This is meant to force compile time check that the enum definition hasn't changed.
+        let a = InvType::Error;
+        match a {
+            InvType::Error => (),
+            InvType::Transaction => (),
+            InvType::Block => (),
+            InvType::WitnessBlock => (),
+            InvType::WitnessTransaction => (),
+        };
+        // TODO: FIXME when #369 is resolved.
+        match rng.gen_range(0, 3) {
+            0 => InvType::Error,
+            1 => InvType::Transaction,
+            2 => InvType::Block,
+//            3 => InvType::WitnessBlock,
+//            4 => InvType::WitnessTransaction,
+            _ => panic!("Please update the random generation for InvType"),
+        }
+    }
+}
+
+impl Distribution<RejectReason> for Standard {
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> RejectReason {
+        // This is meant to force compile time check that the enum definition hasn't changed.
+        let a = RejectReason::Malformed;
+        match a {
+            RejectReason::Malformed => (),
+            RejectReason::Invalid => (),
+            RejectReason::Obsolete => (),
+            RejectReason::Duplicate => (),
+            RejectReason::NonStandard => (),
+            RejectReason::Dust => (),
+            RejectReason::Fee => (),
+            RejectReason::Checkpoint => (),
+        };
+        match rng.gen_range(0, 8) {
+            0 => RejectReason::Malformed,
+            1 => RejectReason::Invalid,
+            2 => RejectReason::Obsolete,
+            3 => RejectReason::Duplicate,
+            4 => RejectReason::NonStandard,
+            5 => RejectReason::Dust,
+            6 => RejectReason::Fee,
+            7 => RejectReason::Checkpoint,
+            _ => panic!("Please update the random generation for RejectReason"),
```

Comment:
can't this just be `unreachable!()`?

Reply:
unreachable!("unreachable can also take a string argument")